lib/madwf_tensor.cu

```diff
@@ -0,0 +1,123 @@
+#include <color_spinor_field.h>
+#include <tunable_nd.h>
+#include <device_vector.h>
+#include <madwf_transfer.h>
+#include <kernels/madwf_tensor.cuh>
+
+namespace quda
+{
+  namespace madwf_ml
+  {
+
+    template <class storage_type, class matrix_type> class tensor_5D_wrapper : public TunableKernel3D
+    {
+      const ColorSpinorField &out;
+      const ColorSpinorField &in;
+      matrix_type *wm_p;
+
+    private:
+      unsigned int sharedBytesPerThread() const
+      {
+        return in.X(4) * color_spin_dim * 2 * sizeof(typename mapper<storage_type>::type) / out.X(4);
+      }
+
+      unsigned int maxSharedBytesPerBlock() const
+      {
+        return maxDynamicSharedBytesPerBlock();
+      }
+
+      unsigned int sharedBytesPerBlock(const TuneParam &) const { return 0; }
+
+      bool tuneGridDim() const { return false; } // Don't tune the grid dimensions.
```

Comment:
This isn't necessary here.  For grid-size tuning, you would use `TunableGridStrideKernel3D`